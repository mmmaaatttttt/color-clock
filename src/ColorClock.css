.ColorClock {
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: white;
  text-shadow: -0.1px -0.1px 1px #333, 0.1px -0.1px 1px #333,
    -0.1px 0.1px 1px #333, 0.1px 0.1px 1px #333;
}

.ColorClock h1 {
  font-size: 10rem;
}

.ColorClock h4 {
  margin-bottom: 1rem;
}

.ColorClock--active {
  justify-content: space-between;
}

.ColorClock--active::before {
  content: "";
  height: 12.4rem;
  visibility: hidden;
}

.ColorClock--icon-wrapper {
  display: flex;
  justify-content: center;
}

.ColorClock--icon {
  color: #2196f3;
  background-color: white;
  border: 2px solid #2196f3;
  padding: 0.7rem;
  margin: -1rem 2rem 0;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  /* font-size: 0.5em; */
  transition: all 0.5s;
}

.ColorClock--icon svg {
  width: 1.25em;
  height: 1.25em;
}

.ColorClock--icon:hover {
  color: #606c76;
  border-color: #606c76;
  cursor: pointer;
}

.ColorClock--early {
  color: #333;
  text-shadow: none;
}

.ColorClock--early p {
  font-weight: bold;
}