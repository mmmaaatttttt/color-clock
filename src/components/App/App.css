.App {
  position: relative;
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column-reverse;
}

@keyframes slidein_about {
  0% {
    opacity: 0;
    transform: translateY(0%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(100%);
  }
}

@keyframes slidein_form_from_clock {
  0% {
    opacity: 0;
    transform: translateY(-100%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(0%);
  }
}

@keyframes slidein_form_from_about {
  0% {
    opacity: 0;
    transform: translateY(100%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(0%);
  }
}

@keyframes slideout_form_to_about {
  0% {
    opacity: 1;
    transform: translateY(0%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(100%);
  }
}

@keyframes slideout_about {
  0% {
    opacity: 1;
    transform: translateY(100%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(0%);
  }
}


@keyframes slideout_form_to_clock {
  0% {
    opacity: 1;
    transform: translateY(0%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(-100%);
  }
}


@keyframes slidein_clock {
  0% {
    opacity: 0;
    transform: translateY(200%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(100%);
  }
}

.page-exit-active .ClockForm {
  animation: slideout_form_to_about 1000ms forwards;
}

.page-enter-active .ClockForm {
  animation: slidein_form_from_about 1000ms forwards;
}

.page-enter-active .About {
  animation: slidein_about 1000ms forwards;
}

.page-exit-active .About  {
  animation: slideout_about 1000ms forwards;
}

.page-enter-active .ClockForm.ClockForm-animate-up {
  animation: slidein_form_from_clock 1000ms forwards;
}

.page-exit-active .ClockForm.ClockForm-animate-up {
  animation: slideout_form_to_clock 1000ms forwards;
}

.page-enter-active .ColorClock { 
  animation: slidein_clock 1000ms forwards;
}
