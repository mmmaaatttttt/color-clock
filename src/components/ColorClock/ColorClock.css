.ColorClock {
  width: 100vw;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: white;
  text-shadow: -0.1px -0.1px 1px #333, 0.1px -0.1px 1px #333,
    -0.1px 0.1px 1px #333, 0.1px 0.1px 1px #333;
  text-align: center;
}

.ColorClock h1 {
  font-size: 10rem;
}

.ColorClock h4 {
  margin-bottom: 1rem;
}

.ColorClock--active {
  justify-content: space-between;
}

.ColorClock--active::before {
  content: "";
  height: 12.4rem;
  visibility: hidden;
}

.ColorClock--icon-wrapper {
  display: flex;
  justify-content: center;
}

.ColorClock--early {
  color: #333;
  text-shadow: none;
}

.ColorClock--early p {
  font-weight: bold;
}