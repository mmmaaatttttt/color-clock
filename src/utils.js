/**
 * Capitalizes the first letter in a string.
 *
 * @param {String} str - String to be capitalized.
 * @returns {String} Returns the capitalized string.
 */
export function capitalize(str) {
  return str[0].toUpperCase() + str.slice(1);
}

/**
 * Converts a time of the form "HH:MM" or "HH:MM:SS"
 * to a number in seconds.
 *
 * @param {String} time - Time in one of the above formats.
 * @returns {Number} Returns a number in seconds.
 */
export function timeToNumber(time) {
  const secondsPerHour = 3600;
  const secondsPerMinute = secondsPerHour / 60;
  let timeArr = time.split(":").map(Number);

  // if no seconds provided, default to 0 seconds
  if (timeArr.length === 2) timeArr.push(0);

  return (
    timeArr[0] * secondsPerHour + timeArr[1] * secondsPerMinute + timeArr[2]
  );
}
