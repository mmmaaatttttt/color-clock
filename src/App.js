import React from 'react';
import ColorClock from "./ColorClock";

function App() {
  return (
    <ColorClock />
  );
}

export default App;
