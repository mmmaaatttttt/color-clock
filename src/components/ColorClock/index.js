import React, { useState, useEffect, useCallback } from "react";
import { useHistory } from "react-router-dom";
import { scaleLinear } from "d3-scale";
import { faEye, faEyeSlash, faRedo } from "@fortawesome/free-solid-svg-icons";
import ProgressBar from "../ProgressBar";
import Icon from "../Icon";
import {
  timeToNumber,
  getTime,
  howLongFromNow,
  middleTimeAsPercentage
} from "../../utils";
import "./ColorClock.css";

function ColorClock({
  times = ["09:00", "10:00", "11:00"],
  colors = ["#00ff00", "#ffff00", "#ff0000"]
}) {
  const history = useHistory();
  const colorScale = useCallback(
    scaleLinear().domain(times.map(timeToNumber)).range(colors).clamp(true),
    [times, colors]
  );
  const [
    { currentTimeForDisplay, currentTimeForComparison },
    setTime
  ] = useState(getTime());
  const [color, setColor] = useState(
    colorScale(timeToNumber(currentTimeForComparison))
  );
  const [showBar, setShowBar] = useState(true);

  const toggleBar = () => {
    setShowBar(barState => !barState);
  };

  const handleReset = () => {
    const msg =
      "Are you sure you want to go back? This action can't be undone.";
    if (window.confirm(msg)) {
      history.push("/", { direction: "up" });
    }
  };

  useEffect(() => {
    const timerId = setInterval(() => {
      const newTime = getTime();
      setTime(newTime);
      setColor(colorScale(timeToNumber(newTime.currentTimeForComparison)));
    }, 1000);

    return () => clearTimeout(timerId);
  }, [colorScale]);

  const startTime = times[0];
  const endTime = times[times.length - 1];

  if (currentTimeForComparison < startTime)
    return (
      <div className="ColorClock ColorClock--early">
        <p>This clock will start {howLongFromNow(startTime)}.</p>
        <p>Current time: {currentTimeForDisplay}.</p>
        <Icon onClick={handleReset} faIcon={faRedo} />
      </div>
    );

  const colorStyle = {
    backgroundColor: color,
    boxShadow: `${color} 0 -5rem 10rem 5rem`
  };

  if (currentTimeForComparison > endTime) {
    return (
      <div className="ColorClock" style={colorStyle}>
        <h1>Time's Up!</h1>
        <h4>This clock ended {howLongFromNow(endTime)}.</h4>
        <h4>Current time: {currentTimeForDisplay}.</h4>
        <Icon onClick={handleReset} faIcon={faRedo} />
      </div>
    );
  }

  return (
    <div className="ColorClock ColorClock--active" style={colorStyle}>
      <div>
        <h1>{currentTimeForDisplay}</h1>
        <div className="ColorClock--icon-wrapper">
          <Icon onClick={toggleBar} faIcon={showBar ? faEyeSlash : faEye} />
          <Icon onClick={handleReset} faIcon={faRedo} />
        </div>
      </div>
      <ProgressBar
        caretPercentage={middleTimeAsPercentage(
          startTime,
          currentTimeForComparison,
          endTime
        )}
        gradientPercentage={middleTimeAsPercentage(...times)}
        colors={colors}
        hidden={!showBar}
      />
    </div>
  );
}

export default React.memo(ColorClock);
