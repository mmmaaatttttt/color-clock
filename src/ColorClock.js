import React, { useState, useEffect } from "react";
import { scaleLinear } from "d3-scale";
import { timeToNumber } from "./utils";
import "./ColorClock.css";

function ColorClock({
  startTime,
  warningTime,
  endTime,
  startColor,
  warningColor,
  endColor
}) {
  const [time, setTime] = useState(new Date());
  const [color, setColor] = useState(startColor);
  const timeStrDisplay = time.toLocaleTimeString("en-US", {
    hour: "2-digit",
    minute: "2-digit"
  });
  const timeStrInternal = time.toLocaleTimeString("en-US", { hour12: false });

  useEffect(() => {
    const timerId = setInterval(() => {
      const newTime = new Date();
      const newTimeStr = newTime.toLocaleTimeString("en-US", { hour12: false });
      const timeArr = [startTime, warningTime, endTime].map(timeToNumber);
      const colorArr = [startColor, warningColor, endColor];
      const colorScale = scaleLinear().domain(timeArr).range(colorArr);
      setTime(new Date());
      setColor(colorScale(timeToNumber(newTimeStr)));
    }, 1000);

    return () => clearTimeout(timerId);
  }, [startTime, startColor, warningTime, warningColor, endTime, endColor]);

  if (timeStrInternal < startTime)
    return (
      <div>
        <p>This clock will start at {startTime}.</p>
        <p>Current time: {timeStrDisplay}.</p>
      </div>
    );

  if (timeStrInternal > endTime) {
    return (
      <div className="ColorClock" style={{ backgroundColor: endColor }}>
        <p>Time's up! This clock ended at {endTime}.</p>
        <p>Current time: {timeStrDisplay}.</p>
      </div>
    );
  }

  return (
    <div className="ColorClock" style={{ backgroundColor: color }}>
      <h1>{timeStrDisplay}</h1>
    </div>
  );
}

export default ColorClock;
