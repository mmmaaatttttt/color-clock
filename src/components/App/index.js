import React from "react";
import Routes from "../Routes";
import "./App.css";

/** 
 * Root of our react application. Related stylesheet includes
 * all keyframes for page transitions.
 * 
 * Props: none
 * State: none
 */
function App() {
  return (
    <div className="App">
      <Routes />
    </div>
  );
}

export default React.memo(App);
