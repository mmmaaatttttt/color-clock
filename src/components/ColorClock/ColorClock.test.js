import React from "react";
import { render, fireEvent } from "@testing-library/react";
import App from "../App";
import ColorClock from "./";
import { confirmMock, timeMock, renderWithRouter } from "../../utils/testUtils";

describe("ColorClock component", function () {
  describe("basic tests", function () {
    it("renders without crashing", function () {
      render(<ColorClock />);
    });

    it("matches snapshot before the clock is active", function () {
      timeMock.set("08:00:00");
      const { asFragment } = render(<ColorClock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot while the clock is active", function () {
      timeMock.set("10:00:00");
      const { asFragment } = render(<ColorClock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot after the clock is active", function () {
      timeMock.set("14:00:00");
      const { asFragment } = render(<ColorClock />);
      expect(asFragment()).toMatchSnapshot();
    });

    afterEach(function () {
      timeMock.reset();
    });
  });

  describe("icon tests - active clock", function () {
    it("clicking on the eye toggles the progress bar", function () {
      timeMock.set("10:00:00");

      const { getAllByRole, getByTestId } = renderWithRouter(<ColorClock />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const progressBar = getByTestId("ProgressBar");
      const eyeIcon = getAllByRole("button")[0];

      expect(eyeIcon.children[0]).toHaveClass("fa-eye-slash");
      expect(eyeIcon.children[0]).not.toHaveClass("fa-eye");
      expect(progressBar).not.toHaveClass("hidden");

      fireEvent.click(eyeIcon);

      // after clicking on the icon, the icon should change,
      // and the progress bar should be hidden
      expect(eyeIcon.children[0]).not.toHaveClass("fa-eye-slash");
      expect(eyeIcon.children[0]).toHaveClass("fa-eye");
      expect(progressBar).toHaveClass("hidden");
    });

    it("clicking on the reset arrow sends you back to the form if you confirm", async function () {
      timeMock.set("10:00:00");

      // mock clicking 'ok' when asked to confirm if you want to go back
      confirmMock.set(true);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[1];

      fireEvent.click(resetIcon);

      // expect the form to appear
      expect(queryByText("10:00 AM")).not.toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).toBeInTheDocument();
    });

    it("clicking on the reset arrow keeps you on the clock if you don't confirm", function () {
      timeMock.set("10:00:00");

      // mock clicking 'cancel' when asked to confirm if you want to go back
      confirmMock.set(false);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[1];

      fireEvent.click(resetIcon);

      // expect the clock to stay, no form to appear
      expect(queryByText("10:00 AM")).toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).not.toBeInTheDocument();
    });

    afterEach(function () {
      timeMock.reset();
      confirmMock.reset();
    });
  });

  describe("icon tests - too early", function () {
    it("clicking on the reset arrow sends you back to the form if you confirm", async function () {
      timeMock.set("06:00:00");

      // mock clicking 'ok' when asked to confirm if you want to go back
      confirmMock.set(true);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[0];

      fireEvent.click(resetIcon);

      // expect the form to appear
      expect(queryByText("Current time: 6:00 AM")).not.toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).toBeInTheDocument();
    });

    it("clicking on the reset arrow keeps you on the clock if you don't confirm", function () {
      timeMock.set("06:00:00");

      // mock clicking 'cancel' when asked to confirm if you want to go back
      confirmMock.set(false);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[0];

      fireEvent.click(resetIcon);

      // expect the clock to stay, no form to appear
      expect(resetIcon).toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).not.toBeInTheDocument();
    });

    afterEach(function () {
      timeMock.reset();
      confirmMock.reset();
    });
  });

  describe("icon tests - too late", function () {
    it("clicking on the reset arrow sends you back to the form if you confirm", async function () {
      timeMock.set("23:00:00");

      // mock clicking 'ok' when asked to confirm if you want to go back
      confirmMock.set(true);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[0];

      fireEvent.click(resetIcon);

      // expect the form to appear
      expect(queryByText("Time's Up!")).not.toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).toBeInTheDocument();
    });

    it("clicking on the reset arrow keeps you on the clock if you don't confirm", function () {
      timeMock.set("23:00:00");

      // mock clicking 'cancel' when asked to confirm if you want to go back
      confirmMock.set(false);

      const { getAllByRole, queryByText } = renderWithRouter(<App />, {
        route: "/clocks/3tH42XskKAkY7NYHrms"
      });

      const resetIcon = getAllByRole("button")[0];

      fireEvent.click(resetIcon);

      // expect the clock to stay, no form to appear
      expect(queryByText("Time's Up!")).toBeInTheDocument();
      expect(
        queryByText("When do you want the clock to start?")
      ).not.toBeInTheDocument();
    });

    afterEach(function () {
      timeMock.reset();
      confirmMock.reset();
    });
  });
});
