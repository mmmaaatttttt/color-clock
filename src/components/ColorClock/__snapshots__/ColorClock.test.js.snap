// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic ColorClock tests matches snapshot after the clock is active 1`] = `
<DocumentFragment>
  <div
    class="ColorClock"
    style="background-color: rgb(255, 0, 0); box-shadow: rgb(255, 0, 0) 0 -5rem 10rem 5rem;"
  >
    <h1>
      Time's Up!
    </h1>
    <h4>
      This clock ended 3 hours ago.
    </h4>
    <h4>
      Current time: 2:00 PM.
    </h4>
    <div
      class="Icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-redo fa-w-16 fa-fw "
        data-icon="redo"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M500.33 0h-47.41a12 12 0 0 0-12 12.57l4 82.76A247.42 247.42 0 0 0 256 8C119.34 8 7.9 119.53 8 256.19 8.1 393.07 119.1 504 256 504a247.1 247.1 0 0 0 166.18-63.91 12 12 0 0 0 .48-17.43l-34-34a12 12 0 0 0-16.38-.55A176 176 0 1 1 402.1 157.8l-101.53-4.87a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12h200.33a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`basic ColorClock tests matches snapshot before the clock is active 1`] = `
<DocumentFragment>
  <div
    class="ColorClock ColorClock--early"
  >
    <p>
      This clock will start in an hour.
    </p>
    <p>
      Current time: 8:00 AM.
    </p>
    <div
      class="Icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-redo fa-w-16 fa-fw "
        data-icon="redo"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M500.33 0h-47.41a12 12 0 0 0-12 12.57l4 82.76A247.42 247.42 0 0 0 256 8C119.34 8 7.9 119.53 8 256.19 8.1 393.07 119.1 504 256 504a247.1 247.1 0 0 0 166.18-63.91 12 12 0 0 0 .48-17.43l-34-34a12 12 0 0 0-16.38-.55A176 176 0 1 1 402.1 157.8l-101.53-4.87a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12h200.33a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`basic ColorClock tests matches snapshot while the clock is active 1`] = `
<DocumentFragment>
  <div
    class="ColorClock ColorClock--active"
    style="background-color: rgb(255, 255, 0); box-shadow: rgb(255, 255, 0) 0 -5rem 10rem 5rem;"
  >
    <div>
      <h1>
        10:00 AM
      </h1>
      <div
        class="ColorClock--icon-wrapper"
      >
        <div
          class="Icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-eye-slash fa-w-20 fa-fw "
            data-icon="eye-slash"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="Icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-redo fa-w-16 fa-fw "
            data-icon="redo"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500.33 0h-47.41a12 12 0 0 0-12 12.57l4 82.76A247.42 247.42 0 0 0 256 8C119.34 8 7.9 119.53 8 256.19 8.1 393.07 119.1 504 256 504a247.1 247.1 0 0 0 166.18-63.91 12 12 0 0 0 .48-17.43l-34-34a12 12 0 0 0-16.38-.55A176 176 0 1 1 402.1 157.8l-101.53-4.87a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12h200.33a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="ProgressBar "
    >
      <div
        class="ProgressBar--bar"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-up fa-w-10 fa-2x ProgressBar--caret"
          data-icon="caret-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="left: 50%;"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
